language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/ModVarToListenerDeclCodeAction.java

```diff
@@ -0,0 +1,137 @@
+import io.ballerina.tools.diagnostics.Diagnostic;
+import io.ballerina.tools.diagnostics.Location;
+import org.apache.commons.lang3.tuple.Pair;
+import org.ballerinalang.annotation.JavaSPIService;
+import org.ballerinalang.langserver.common.constants.CommandConstants;
+import org.ballerinalang.langserver.common.utils.CommonUtil;
+import org.ballerinalang.langserver.commons.CodeActionContext;
+import org.ballerinalang.langserver.commons.codeaction.spi.DiagBasedPositionDetails;
+import org.ballerinalang.util.diagnostic.DiagnosticErrorCode;
+import org.eclipse.lsp4j.CodeAction;
+import org.eclipse.lsp4j.Position;
+import org.eclipse.lsp4j.Range;
+import org.eclipse.lsp4j.TextEdit;
+
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+
+/**
+ * Code action for changing a module var declaration statement to listener declaration.
+ *
+ * @since 2.0.0
+ */
+@JavaSPIService("org.ballerinalang.langserver.commons.codeaction.spi.LSCodeActionProvider")
+public class ModVarToListenerDeclCodeAction extends AbstractCodeActionProvider {
+
+    public static final String NAME = "Module var to listener declaration";
+
+    @Override
+    public List<CodeAction> getDiagBasedCodeActions(Diagnostic diagnostic,
+                                                    DiagBasedPositionDetails positionDetails,
+                                                    CodeActionContext context) {
+        if (!(DiagnosticErrorCode.INVALID_LISTENER_ATTACHMENT.diagnosticId()
+                .equals(diagnostic.diagnosticInfo().code()))) {
+            return Collections.emptyList();
+        }
+
+        Node matchedNode = positionDetails.matchedNode();
+        if (matchedNode == null) {
+            return Collections.emptyList();
+        }
+        Optional<Pair<CaptureBindingPatternNode, String>> captureBindingPatternNodeFilePathPair =
+                findCaptureBindingPattern(matchedNode, context);
+        if (captureBindingPatternNodeFilePathPair.isEmpty() ||
+                captureBindingPatternNodeFilePathPair.get().getLeft().parent().kind()
+                        != SyntaxKind.TYPED_BINDING_PATTERN) {
+            return Collections.emptyList();
+        }
+        TypedBindingPatternNode typedBindingPatternNode =
+                (TypedBindingPatternNode) captureBindingPatternNodeFilePathPair.get().getLeft().parent();
+        List<CodeAction> actions = new ArrayList<>();
+        List<TextEdit> textEdits = new ArrayList<>();
+        Position pos = CommonUtil.toRange(typedBindingPatternNode.lineRange()).getStart();
+        Position insertPos = new Position(pos.getLine(), pos.getCharacter());
+        textEdits.add(new TextEdit(new Range(insertPos, insertPos),
+                SyntaxKind.LISTENER_KEYWORD.stringValue() + " "));
+        String commandTitle = String.format(CommandConstants.CONVERT_MODULE_VAR_TO_LISTENER_DECLARATION,
+                matchedNode.toSourceCode());
```

Comment:
when we get the value from the `sourceCode` API for displaying purposes, make sure you trim the value.